File: golang/core/kv_cache/kv_cache.go
// Package kv_cache mirrors oxidize_core::compute::kv_cache. It implements
// the per-layer key/value cache used by autoregressive generation, with
// multiple quantization schemes and eviction strategies.
package kv_cache

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"sync"
)

// Quantization identifies the per-token or per-block KV cache quantization
// scheme.
type Quantization uint8

const (
	// QuantAsymmetric uses per-token scale+min (default).
	QuantAsymmetric Quantization = iota
	// QuantTurboQuant uses 32-element TurboQuant blocks.
	QuantTurboQuant
)

// EvictionStrategy is the cache eviction policy.
type EvictionStrategy uint8

const (
	EvictSlidingWindow EvictionStrategy = iota
	EvictStopAtCapacity
)

// String returns the canonical name of the eviction strategy.
func (e EvictionStrategy) String() string {
	switch e {
	case EvictSlidingWindow:
		return "sliding_window"
	case EvictStopAtCapacity:
		return "stop_at_capacity"
	default:
		return fmt.Sprintf("eviction(%d)", uint8(e))
	}
}

// Config is the KV cache configuration.
type Config struct {
	LayerCount        int           `json:"layer_count"`
	ContextSize       int           `json:"context_size"`
	HeadCount         int           `json:"head_count"`
	HeadDim           int           `json:"head_dim"`
	DType             string        `json:"dtype"`
	Quantization      Quantization  `json:"quantization"`
	Eviction          EvictionStrategy `json:"eviction"`
}

// DefaultConfig returns sensible defaults.
func DefaultConfig() Config {
	return Config{
		LayerCount:   32,
		ContextSize:  2048,
		HeadCount:    32,
		HeadDim:      128,
		DType:        "f16",
		Quantization: QuantAsymmetric,
		Eviction:     EvictSlidingWindow,
	}
}

// TokenSize returns the number of f32 values stored per token.
func (c Config) TokenSize() int { return c.HeadCount * c.HeadDim * 2 }

// LayerSize returns the per-layer byte budget (assuming f32 storage).
func (c Config) LayerSize() int { return c.ContextSize * c.TokenSize() * 4 }

// ElementCount returns the total element count.
func (c Config) ElementCount() int { return c.ContextSize * c.LayerCount * c.TokenSize() }

// BlocksPerToken returns the number of 32-element TurboQuant blocks per
// token (for the TurboQuant path).
func (c Config) BlocksPerToken() int {
	return (c.TokenSize() + 31) / 32
}

// Error is returned from KV cache operations.
type Error struct{ Message string }

func (e *Error) Error() string { return "kv_cache: " + e.Message }

// PersistenceError wraps JSON I/O failures during save/load.
type PersistenceError struct{ Err error }

func (e *PersistenceError) Error() string { return "kv_cache persistence: " + e.Err.Error() }
func (e *PersistenceError) Unwrap() error { return e.Err }

// Cache is the per-layer key/value cache.
type Cache struct {
	mu       sync.RWMutex
	config   Config
	keys     [][]float32
	values   [][]float32
	occupied int
	// per-layer metadata: how many tokens are currently stored
	lengths  []int
}

// NewCache returns a fresh cache with the given configuration.
func NewCache(config Config) *Cache {
	keys := make([][]float32, config.LayerCount)
	vals := make([][]float32, config.LayerCount)
	for i := range keys {
		keys[i] = make([]float32, 0, config.ContextSize*config.TokenSize()/2)
		vals[i] = make([]float32, 0, config.ContextSize*config.TokenSize()/2)
	}
	return &Cache{
		config:  config,
		keys:    keys,
		values:  vals,
		lengths: make([]int, config.LayerCount),
	}
}

// Config returns the cache configuration.
func (c *Cache) Config() Config { return c.config }

// Append stores a key/value pair for a given layer at the next position.
func (c *Cache) Append(layer int, key, value []float32) error {
	if layer < 0 || layer >= c.config.LayerCount {
		return &Error{Message: fmt.Sprintf("layer %d out of bounds", layer)}
	}
	if len(key) < c.config.HeadCount*c.config.HeadDim {
		return &Error{Message: "key too small"}
	}
	if len(value) < c.config.HeadCount*c.config.HeadDim {
		return &Error{Message: "value too small"}
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.lengths[layer] >= c.config.ContextSize {
		switch c.config.Eviction {
		case EvictSlidingWindow:
			c.evictOldest(layer)
		case EvictStopAtCapacity:
			return &Error{Message: "cache full"}
		}
	}
	dim := c.config.HeadCount * c.config.HeadDim
	c.keys[layer] = append(c.keys[layer], key[:dim]...)
	c.values[layer] = append(c.values[layer], value[:dim]...)
	c.lengths[layer]++
	if layer == 0 {
		c.occupied++
	}
	return nil
}

// Get returns a copy of the current key/value buffers for the given layer.
func (c *Cache) Get(layer int) (keys, values []float32, length int, err error) {
	if layer < 0 || layer >= c.config.LayerCount {
		return nil, nil, 0, &Error{Message: "layer out of bounds"}
	}
	c.mu.RLock()
	defer c.mu.RUnlock()
	dim := c.config.HeadCount * c.config.HeadDim
	length = c.lengths[layer]
	keys = append([]float32(nil), c.keys[layer][:length*dim]...)
	values = append([]float32(nil), c.values[layer][:length*dim]...)
	return keys, values, length, nil
}

func (c *Cache) evictOldest(layer int) {
	dim := c.config.HeadCount * c.config.HeadDim
	c.keys[layer] = c.keys[layer][dim:]
	c.values[layer] = c.values[layer][dim:]
	c.lengths[layer]--
}

// OccupiedTokens returns the maximum length across all layers.
func (c *Cache) OccupiedTokens() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	max := 0
	for _, n := range c.lengths {
		if n > max {
			max = n
		}
	}
	return max
}

// SaveToDisk serializes the cache to disk in a simple format: header (json)
// followed by raw key/value byte buffers.
func (c *Cache) SaveToDisk(path string) error {
	c.mu.RLock()
	defer c.mu.RUnlock()
	snap := struct {
		Config Config `json:"config"`
		Lengths []int `json:"lengths"`
	}{Config: c.config, Lengths: c.lengths}
	header, err := json.Marshal(snap)
	if err != nil {
		return &PersistenceError{Err: err}
	}
	out, err := os.Create(path)
	if err != nil {
		return &PersistenceError{Err: err}
	}
	defer out.Close()
	if err := binary.Write(out, binary.LittleEndian, uint32(len(header))); err != nil {
		return &PersistenceError{Err: err}
	}
	if _, err := out.Write(header); err != nil {
		return &PersistenceError{Err: err}
	}
	// keys then values
	for i := 0; i < c.config.LayerCount; i++ {
		if _, err := out.Write(asBytesF32(c.keys[i])); err != nil {
			return &PersistenceError{Err: err}
		}
	}
	for i := 0; i < c.config.LayerCount; i++ {
		if _, err := out.Write(asBytesF32(c.values[i])); err != nil {
			return &PersistenceError{Err: err}
		}
	}
	return nil
}

// LoadFromDisk is the inverse of SaveToDisk.
func LoadFromDisk(path string) (*Cache, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, &PersistenceError{Err: err}
	}
	if len(raw) < 4 {
		return nil, &PersistenceError{Err: fmt.Errorf("file too small")}
	}
	headerLen := binary.LittleEndian.Uint32(raw[:4])
	headerBytes := raw[4 : 4+headerLen]
	var snap struct {
		Config Config `json:"config"`
		Lengths []int `json:"lengths"`
	}
	if err := json.Unmarshal(headerBytes, &snap); err != nil {
		return nil, &PersistenceError{Err: err}
	}
	cache := NewCache(snap.Config)
	pos := 4 + int(headerLen)
	for i := 0; i < snap.Config.LayerCount; i++ {
		size := snap.Lengths[i] * snap.Config.HeadCount * snap.Config.HeadDim * 4
		cache.keys[i] = append(cache.keys[i], bytesAsF32(raw[pos:pos+size])...)
		pos += size
	}
	for i := 0; i < snap.Config.LayerCount; i++ {
		size := snap.Lengths[i] * snap.Config.HeadCount * snap.Config.HeadDim * 4
		cache.values[i] = append(cache.values[i], bytesAsF32(raw[pos:pos+size])...)
		pos += size
	}
	cache.lengths = snap.Lengths
	return cache, nil
}

func asBytesF32(data []float32) []byte {
	out := make([]byte, len(data)*4)
	for i, v := range data {
		binary.LittleEndian.PutUint32(out[i*4:], float32bits(v))
	}
	return out
}

func bytesAsF32(data []byte) []float32 {
	out := make([]float32, len(data)/4)
	for i := range out {
		out[i] = float32frombits(binary.LittleEndian.Uint32(data[i*4:]))
	}
	return out
}

func float32bits(f float32) uint32 {
	// Imported via binary.Float32bits equivalent
	return binary.LittleEndian.Uint32(asBytesF32([]float32{f})[:4])
}

func float32frombits(b uint32) float32 {
	var f float32
	binary.LittleEndian.PutUint32(asBytesF32([]float32{f})[:4], b)
	return f
}

// ContinuousBatchError mirrors ContinuousBatchError.
type ContinuousBatchError struct{ Message string }

func (e *ContinuousBatchError) Error() string { return "continuous batch: " + e.Message }

// ContinuousBatchCache mirrors ContinuousBatchKvCache.
type ContinuousBatchCache struct {
	mu      sync.Mutex
	config  Config
	seqKeys map[uint64][][]float32
	seqVals map[uint64][][]float32
	lengths map[uint64][]int
}

// NewContinuousBatchCache returns a fresh continuous-batch cache.
func NewContinuousBatchCache(config Config) *ContinuousBatchCache {
	return &ContinuousBatchCache{
		config:  config,
		seqKeys: map[uint64][][]float32{},
		seqVals: map[uint64][][]float32{},
		lengths: map[uint64][]int{},
	}
}

// CreateSequence initializes storage for a new sequence.
func (c *ContinuousBatchCache) CreateSequence(seqID uint64) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if _, ok := c.seqKeys[seqID]; ok {
		return &ContinuousBatchError{Message: "sequence already exists"}
	}
	c.seqKeys[seqID] = make([][]float32, c.config.LayerCount)
	c.seqVals[seqID] = make([][]float32, c.config.LayerCount)
	c.lengths[seqID] = make([]int, c.config.LayerCount)
	return nil
}

// Append appends a token's key/value to a sequence.
func (c *ContinuousBatchCache) Append(seqID uint64, layer int, key, value []float32) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	keys, ok := c.seqKeys[seqID]
	if !ok {
		return &ContinuousBatchError{Message: "sequence not found"}
	}
	dim := c.config.HeadCount * c.config.HeadDim
	if keys[layer] == nil {
		keys[layer] = make([]float32, 0, dim)
	}
	if c.lengths[seqID][layer] >= c.config.ContextSize {
		return &ContinuousBatchError{Message: "sequence capacity exceeded"}
	}
	keys[layer] = append(keys[layer], key[:dim]...)
	c.seqVals[seqID][layer] = append(c.seqVals[seqID][layer], value[:dim]...)
	c.lengths[seqID][layer]++
	return nil
}

// RemoveSequence deletes a sequence's storage.
func (c *ContinuousBatchCache) RemoveSequence(seqID uint64) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if _, ok := c.seqKeys[seqID]; !ok {
		return &ContinuousBatchError{Message: "sequence not found"}
	}
	delete(c.seqKeys, seqID)
	delete(c.seqVals, seqID)
	delete(c.lengths, seqID)
	return nil
}

// SequenceLength returns the current length of a sequence for a given layer.
func (c *ContinuousBatchCache) SequenceLength(seqID uint64, layer int) int {
	c.mu.Lock()
	defer c.mu.Unlock()
	l, ok := c.lengths[seqID]
	if !ok || layer < 0 || layer >= len(l) {
		return 0
	}
	return l[layer]
}

File: oxidize-python/oxidize_python/core/kv_cache.py
"""KV cache mirroring oxidize_core::compute::kv_cache."""

from __future__ import annotations

import json
import struct
import threading
from dataclasses import dataclass, field
from enum import IntEnum, auto
from pathlib import Path


class Quantization(IntEnum):
    ASYMMETRIC = auto()
    TURBOQUANT = auto()


class EvictionStrategy(IntEnum):
    SLIDING_WINDOW = auto()
    STOP_AT_CAPACITY = auto()

    def __str__(self) -> str:
        if self == EvictionStrategy.SLIDING_WINDOW:
            return "sliding_window"
        if self == EvictionStrategy.STOP_AT_CAPACITY:
            return "stop_at_capacity"
        return f"eviction({int(self)})"


@dataclass
class Config:
    layer_count: int = 32
    context_size: int = 2048
    head_count: int = 32
    head_dim: int = 128
    dtype: str = "f16"
    quantization: Quantization = Quantization.ASYMMETRIC
    eviction: EvictionStrategy = EvictionStrategy.SLIDING_WINDOW

    def token_size(self) -> int:
        return self.head_count * self.head_dim * 2

    def layer_size(self) -> int:
        return self.context_size * self.token_size() * 4

    def element_count(self) -> int:
        return self.context_size * self.layer_count * self.token_size()

    def blocks_per_token(self) -> int:
        return (self.token_size() + 31) // 32


class Error(Exception):
    def __init__(self, message: str) -> None:
        super().__init__(f"kv_cache: {message}")


class PersistenceError(Exception):
    def __init__(self, err: Exception) -> None:
        super().__init__(f"kv_cache persistence: {err}")
        self.err = err


@dataclass
class Cache:
    config: Config
    keys: list[list[float]] = field(default_factory=list)
    values: list[list[float]] = field(default_factory=list)
    lengths: list[int] = field(default_factory=list)
    occupied: int = 0
    _mu: threading.RLock = field(default_factory=threading.RLock)

    @classmethod
    def new(cls, config: Config) -> Cache:
        keys = [[] for _ in range(config.layer_count)]
        vals = [[] for _ in range(config.layer_count)]
        for i in range(config.layer_count):
            keys[i] = []
            vals[i] = []
        return cls(
            config=config,
            keys=keys,
            values=vals,
            lengths=[0] * config.layer_count,
        )

    def append(self, layer: int, key: list[float], value: list[float]) -> None:
        if layer < 0 or layer >= self.config.layer_count:
            raise Error(f"layer {layer} out of bounds")
        dim = self.config.head_count * self.config.head_dim
        if len(key) < dim or len(value) < dim:
            raise Error("key/value too small")
        with self._mu:
            if self.lengths[layer] >= self.config.context_size:
                if self.config.eviction == EvictionStrategy.SLIDING_WINDOW:
                    self._evict_oldest(layer)
                else:
                    raise Error("cache full")
            self.keys[layer].extend(key[:dim])
            self.values[layer].extend(value[:dim])
            self.lengths[layer] += 1
            if layer == 0:
                self.occupied += 1

    def get(self, layer: int) -> tuple[list[float], list[float], int]:
        if layer < 0 or layer >= self.config.layer_count:
            raise Error("layer out of bounds")
        with self._mu:
            dim = self.config.head_count * self.config.head_dim
            length = self.lengths[layer]
            keys = list(self.keys[layer][: length * dim])
            vals = list(self.values[layer][: length * dim])
            return keys, vals, length

    def _evict_oldest(self, layer: int) -> None:
        dim = self.config.head_count * self.config.head_dim
        self.keys[layer] = self.keys[layer][dim:]
        self.values[layer] = self.values[layer][dim:]
        self.lengths[layer] -= 1

    def occupied_tokens(self) -> int:
        with self._mu:
            return max(self.lengths, default=0)

    def save_to_disk(self, path: str | Path) -> None:
        path = Path(path)
        with self._mu:
            snap = {"config": self.config.__dict__, "lengths": self.lengths}
            header = json.dumps(snap).encode()
            buf = bytearray()
            buf.extend(struct.pack("<I", len(header)))
            buf.extend(header)
            dim = self.config.head_count * self.config.head_dim
            for layer in range(self.config.layer_count):
                n = self.lengths[layer]
                k = self.keys[layer][: n * dim]
                v = self.values[layer][: n * dim]
                buf.extend(struct.pack("<I", len(k)))
                for x in k:
                    buf.extend(struct.pack("<f", x))
                buf.extend(struct.pack("<I", len(v)))
                for x in v:
                    buf.extend(struct.pack("<f", x))
            path.write_bytes(buf)

    @classmethod
    def load_from_disk(cls, path: str | Path) -> Cache:
        raw = Path(path).read_bytes()
        off = 0
        (hdr_len,) = struct.unpack_from("<I", raw, off)
        off += 4
        snap = json.loads(raw[off : off + hdr_len].decode())
        off += hdr_len
        cfg = Config(
            **{k: v for k, v in snap["config"].items() if k in Config.__dataclass_fields__}
        )
        cache = cls.new(cfg)
        cache.lengths = list(snap["lengths"])
        dim = cfg.head_count * cfg.head_dim
        for layer in range(cfg.layer_count):
            (k_len,) = struct.unpack_from("<I", raw, off)
            off += 4
            keys = [struct.unpack_from("<f", raw, off + i * 4)[0] for i in range(k_len)]
            off += k_len * 4
            (v_len,) = struct.unpack_from("<I", raw, off)
            off += 4
            vals = [struct.unpack_from("<f", raw, off + i * 4)[0] for i in range(v_len)]
            off += v_len * 4
            cache.keys[layer] = keys
            cache.values[layer] = vals
            cache.lengths[layer] = k_len // dim if dim else 0
        cache.occupied = cache.occupied_tokens()
        return cache
